// nextra/dist/server/utils.js
import title from "title";
import { DEFAULT_PROPERTY_PROPS } from "./constants.js";
const logger = {
  info: console.log.bind(null, "-", "\x1B[36minfo\x1B[0m", "[nextra]"),
  warn: console.log.bind(null, "-", "\x1B[33mwarn\x1B[0m", "[nextra]"),
  error: console.log.bind(null, "-", "\x1B[31merror\x1B[0m", "[nextra]")
};
function pageTitleFromFilename(fileName) {
  return title(fileName.replaceAll(/[-_]/g, " "), {
    special: ["SSR", "CORS", "ESLint"]
  });
}
function createAstExportConst(name, value) {
  return {
    type: "ExportNamedDeclaration",
    specifiers: [],
    declaration: {
      type: "VariableDeclaration",
      kind: "const",
      declarations: [
        {
          type: "VariableDeclarator",
          id: { type: "Identifier", name },
          init: value
        }
      ]
    }
  };
}
function createAstObject(obj) {
  return {
    type: "ObjectExpression",
    properties: Object.entries(obj).map(([key, value]) => ({
      ...DEFAULT_PROPERTY_PROPS,
      key: { type: "Identifier", name: key },
      value: value && typeof value === "object" ? value : { type: "Literal", value }
    }))
  };
}
export {
  createAstExportConst,
  createAstObject,
  logger,
  pageTitleFromFilename
};
