// nextra/dist/server/page-map/to-ast.js
import path from "path";
import { MARKDOWN_EXTENSION_RE } from "../constants.js";
import { createAstObject } from "../utils.js";
function cleanFilePath(filePath) {
  const { dir, name } = path.parse(filePath);
  const fileName = `${dir.replace(/^(src\/)?content\/?/, "")}_${name}`;
  return (
    // handle non-"\w" characters
    encodeURI(fileName).replaceAll(/[\W_]+/g, "_").replace(/^_/, "").replace(/^\d/, (match) => `_${match}`)
  );
}
function convertPageMapToAst(pageMap, imports) {
  const elements = pageMap.map((item) => {
    if ("children" in item) {
      return createAstObject({
        name: item.name,
        route: item.route,
        children: convertPageMapToAst(item.children, imports)
      });
    }
    if ("route" in item) {
      const filePath2 = item.__pagePath;
      const importName2 = cleanFilePath(filePath2);
      imports.push({ importName: importName2, filePath: filePath2 });
      return createAstObject({
        name: item.name,
        route: item.route,
        frontMatter: MARKDOWN_EXTENSION_RE.test(filePath2) ? { type: "Identifier", name: importName2 } : {
          type: "CallExpression",
          callee: { type: "Identifier", name: "getMetadata" },
          arguments: [{ type: "Identifier", name: importName2 }],
          optional: false
        }
      });
    }
    const filePath = item.__metaPath;
    const importName = cleanFilePath(filePath);
    imports.push({ importName, filePath });
    return createAstObject({
      data: { type: "Identifier", name: importName }
    });
  });
  return { type: "ArrayExpression", elements };
}
export {
  convertPageMapToAst
};
