// nextra/dist/server/rehype-plugins/rehype-twoslash-popup.js
import { EXIT, visit } from "unist-util-visit";
const TWOSLASH_POPUP_IMPORT_AST = {
  type: "mdxjsEsm",
  data: {
    estree: {
      body: [
        {
          type: "ImportDeclaration",
          source: { type: "Literal", value: "nextra/components" },
          specifiers: [
            {
              type: "ImportSpecifier",
              imported: { type: "Identifier", name: "Popup" },
              local: { type: "Identifier", name: "Popup" }
            }
          ]
        }
      ]
    }
  }
};
const transformer = (ast) => {
  visit(
    ast,
    [
      { tagName: "popup" },
      { tagName: "popupbutton" },
      { tagName: "popuppanel" }
    ],
    (node) => {
      const n = node;
      const tagName = {
        popup: "Popup",
        popupbutton: "Popup.Button",
        popuppanel: "Popup.Panel"
      }[n.tagName];
      n.tagName = tagName;
    }
  );
  visit(ast, { tagName: "code" }, (node) => {
    if (node.data?.meta === "twoslash") {
      ast.children.unshift(TWOSLASH_POPUP_IMPORT_AST);
      return EXIT;
    }
  });
};
const rehypeTwoslashPopup = () => transformer;
export {
  rehypeTwoslashPopup
};
